channeldb/payments.go

```diff
@@ -443,6 +443,154 @@ func fetchPayment(bucket *bbolt.Bucket) (*MPPayment, error) {
+	// as well as failed payments will show up if this
+	// field is set to True.
+	IncludeIncomplete bool
+}
+
+// PaymentsQuerySlice contains the result of a query to the payments database.
+// It includes the original query, the set of payments that match the query,
+// and integers which represent the offset index of the first and last item
+// returned in the series of payments. These integers allow callers to resume
+// their query in the event that the query's response exceeds the max number of
+// returnable events.
+type PaymentsQuerySlice struct {
+	PaymentsQuery
+
+	// Payments is the set of payments returned from the database for the
+	// PaymentsQuery.
+	Payments []MPPayment
+
+	// FirstIndexOffset is the index of the first element in the set of
+	// returned MPPayments. Callers can use this to resume their query
+	// in the event that the slice has too many events to fit into a single
+	// response.
+	FirstIndexOffset uint64
+
+	// LastIndexOffset is the index of the last element in the set of
+	// returned MPPayments. Callers can use this to resume their query
+	// in the event that the slice has too many events to fit into a single
+	// response.
+	LastIndexOffset uint64
+}
+
+// startIndex determines the index from where we loop through the payments.
+// indexOffset is excluded and we start at the next index for normal order
+// and at the previous index for reversed order.
+// startIndex returns a boolean, which tells about whether the indexOffset
+// would give rise to payments found in the next when looping through them.
+func startIndex(indexOffset, paymentsLength uint64, reversed bool) (uint64, bool) {
+	if reversed {
+		switch {
+		// When indexOffset is zero or the indexOffset lies outside
+		// of the payments range, we start from the last known payment.
+		case indexOffset == 0 || indexOffset > paymentsLength:
+			return paymentsLength, true
+		case indexOffset > 1:
+			return indexOffset - 1, true
+		default:
+			return 0, false
+		}
+	}
+	// Normal ordering
+	switch {
+	case indexOffset < paymentsLength:
+		return indexOffset + 1, true
+	default:
+		return 0, false
+	}
+}
+
+// nextIndex determines the next payment index in the loop.
+// It includes a boolean which tells about whether there was a next index.
+func nextIndex(index, paymentsLength uint64, reversed bool) (uint64, bool) {
+	switch {
+	case reversed && index > 1:
+		return index - 1, true
+	case !reversed && index < paymentsLength:
+		return index + 1, true
+	// Stop if payments are exhausted.
+	default:
+		return 0, false
+	}
+}
+
+// QueryPayments is a query to the payments database which is restricted
+// to a subset of payments by the payments query, containing an offset
+// index and a maximum number of returned payments.
+func (db *DB) QueryPayments(query PaymentsQuery) (PaymentsQuerySlice, error) {
+	resp := PaymentsQuerySlice{
+		PaymentsQuery: query,
+	}
+	allPayments, err := db.FetchPayments()
```

Comment:
I checked out `FetchPayments` quickly and it looks like the list that is returned here, is sorted by payment hash. That means that if a new payment is initiated, the ordering will be different and everything shifts?

Reply:
No, it's sorted by the sequence number, the equivalent to the add index in the invoices db, so no reordering should occur.

Reply:
Ah right, there is a sort at the end of `FetchPayments`